### ISSUE TITLE
SageMaker endpoint can't load huggingface tokenizer

### ISSUE THREAD

**Issue by elozano98** [aws/sagemaker-huggingface-inference-toolkit#31]
I used Amazon SageMaker to train a HuggingFace model. At the end of the training script provided to the estimator, I saved the model into the correct path (`SM_MODEL_DIR`):
```
if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--model-dir", type=str, default=os.environ["SM_MODEL_DIR"])
    ...
    trainer.model.save_pretrained(args.model_dir)
```
After the model was trained, I deployed it using the `deploy` method of the HuggingFace estimator. Once the endpoint was successfully created, I tried inference with the returned predictor:
```
response = self.predictor.predict(
    {"inputs": "I want to know where is my order"}
)
```
And I received the following client error:
```
{'code': 400, 'type': 'InternalServerException', 'message': "Can't load tokenizer for '/.sagemaker/mms/models/model'. Make sure that:\n\n- '/.sagemaker/mms/models/model' is a correct model identifier listed on 'https://huggingface.co/models'\n\n- or '/.sagemaker/mms/models/model' is the correct path to a directory containing relevant tokenizer files\n\n"}
```

The problem seems to be on the path that the endpoint uses to load the model on the `from_pretrained` method.

Any idea of why the tokenizer cannot be loaded?

**Comment by philschmid** [comment]
Do you save your `tokenizer` as well in your training script? if not this would explain why it cannot be found when deploying the model.

**Comment by elozano98** [comment]
Yes, I've tried saving the tokenizer too, and still get the same message:

```python
if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--model-dir", type=str, default=os.environ["SM_MODEL_DIR"])
    ...
    trainer.model.save_pretrained(args.model_dir)
    trainer.tokenizer.save_pretrained(args.model_dir)
```

**Comment by philschmid** [comment]
Can you please share the structure for your model.tar.gz?

**Comment by elozano98** [comment]
The `model.tar.gz` structure is:

- checkpoint-43
- config.json
- pytorch_model.bin
- special_tokens_map.json
- tokenizer_config.json
- vocab.txt

**Comment by philschmid** [comment]
This should work! after adding `trainer.tokenizer.save_pretrained(args.model_dir)` are you still seeing the same issue?

**Comment by elozano98** [comment]
Yes, I still have the same issue. 

I think it can maybe be something related to the path used by the endpoint when it tries to load the tokenizer using `from_pretrained`. 

Also, all the code that creates the estimator, trains it, deploys it, and uses the predictor for inference, is executed in my local machine. Despite the model is trained, stored, and deployed successfully, maybe this can be related to an error in the path used in the endpoint.

**Comment by philschmid** [comment]
What do you mean with 

> Also, all the code that creates the estimator, trains it, deploys it, and uses the predictor for inference, is executed in my local machine.

Aren't you running the training on AWS? Is the `model.tar.gz` uploaded to s3?

**Comment by elozano98** [comment]
Yes, the model is stored in the S3 bucket. The only difference is that I don't use a notebook instance in SageMaker to create, train and deploy the estimator. But that should work too, since I make sure the model is store in the correct S3 bucket.

**Comment by philschmid** [comment]
Yeah this shouldn't be an issue. Can you try creating a new `model.tar.gz` manually? 
https://huggingface.co/docs/sagemaker/inference#creating-a-model-artifact-modeltargz-for-deployment

Replace step 1 with unzipping your current archive. You can also remove your checkpoint then.

**Comment by segments-tobias** [comment]
@elozano98 Did you ever find the cause of the problem? I might be facing something similar
```
ModelError: An error occurred (ModelError) when calling the InvokeEndpoint operation: Received client error (400) from primary with message "{
  "code": 400,
  "type": "InternalServerException",
  "message": "Can\u0027t load config for \u0027/.sagemaker/mms/models/model\u0027. Make sure that:\n\n- \u0027/.sagemaker/mms/models/model\u0027 is a correct model identifier listed on \u0027https://huggingface.co/models\u0027\n  (make sure \u0027/.sagemaker/mms/models/model\u0027 is not a path to a local directory with something else, in that case)\n\n- or \u0027/.sagemaker/mms/models/model\u0027 is the correct path to a directory containing a config.json file\n\n"
}
```

**Comment by philschmid** [comment]
@segments-tobias could you describe how you have created your `model.tar.gz`?

**Comment by segments-tobias** [comment]
@philschmid I found the answer in another issue. I also accidentally zipped the directory, instead of just the contents.
